### PR TITLE
Implement DataSource's new methods: hasList, hasNested, getListOf, and toMap

### PR DESCRIPTION
**Pull request by dmikurube**
Some methods have been added in Embulk SPI's DataSource: https://github.com/embulk/embulk/pull/1534

It implements them in embulk-util-config's DataSourceImpl.

### CODE REVIEW

**Comment by dmikurube** [comment]
We're going to add unit tests for them.

File: src/test/java/org/embulk/util/config/TestDataSourceImpl.java [hunk 1]
```diff
@@ -40,18 +53,157 @@ public void testGetList() {
...
+            impl.getListOf(String.class, "array");
+        } catch (final ConfigException ex) {
+            assertTrue(ex.getCause() instanceof com.fasterxml.jackson.core.JsonProcessingException);
+            return;
+        }
+        fail("ConfigException should be thrown by getting a mixed JSON array as a List of Strings.");
+    }
+
     @Test
     public void testFailToGetNestedAsList() {
```
Comment (hiroyuki-sato) [inline src/test/java/org/embulk/util/config/TestDataSourceImpl.java:208]: Just memo reading review

```json
{
  "string": "foo",
  "boolean": true,
  "int": 12,
  "double": 42914.142,
  "array": [
    "hoge",
    "fuga",
    {
      "subkey1": "something"
    },
    [
      "sumewhat"
    ]
  ],
  "arrayOfStrings": [
    "foo",
    "bar",
    "baz"
  ],
  "arrayOfIntegers": [
    124,
    -4014,
    9241
  ],
  "object": {
    "key1": "value1",
    "key2": "value2"
  }
}
```

**Review by hiroyuki-sato — APPROVED** [review]
LGTM👍

**Comment by dmikurube** [comment]
Thanks!

**Pull request merged**